### ISSUE TITLE
spec-decode: eval_draft.py always proposes random tokens instead of running the loaded draft model; C++ path still wired to a checkpoint already proven non-functional while a newer, unevaluated one sits unused

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#115]
## Summary

Two related findings in the speculative-decoding pipeline: an evaluation script that silently fabricates its core measurement, and the real C++ inference path pointing at a stale checkpoint even after a newer training run appears to have completed.

## Finding 1: `eval_draft.py` never calls the draft model it loads

`spec-decode/eval/eval_draft.py:150-156`:
```python
draft_tokens = []
for i in range(block_size):
    # In production: draft_model.forward(hidden_states, last_token)
    # For now, simulate with random tokens
    draft_tokens.append(torch.randint(0, 100, (1,), device=device))
```
`draft_model` is loaded (line 288) and threaded through the whole script, but `speculative_decode_greedy()` never calls it — it always proposes uniform-random tokens from a 100-token bucket regardless of checkpoint quality, guaranteeing near-0% acceptance independent of the model. This is presented with no caveat as the official evaluation step in `spec-decode/README.md`'s Quick Start (`### 2. Evaluate — python eval/eval_draft.py ...`). This is the same "placeholder pretending to be real" pattern the repo has already caught and fixed elsewhere (`9791b8d0` "stop emitting fabricated logprobs", `375dc6f4` "correct the DSpark disproven conclusion").

**Mitigating context**: this script appears to be dead/unused — the 0.8 tok/s / 0% acceptance figures actually published in `docs/wiki/performance.md` / `site/benchmarks.json` come from the real C++ path (`tools/dspark_gpu_bench.cpp`, `spec-decode/engine/npu_spec_integration.cpp`), not this script, so current published claims aren't corrupted by it. But it's a landmine: anyone following the README's own Quick Start gets a meaningless number that looks like a real measurement.

## Finding 2: wired-in checkpoint is stale; a newer one exists but was never evaluated or wired in

`tools/dspark_gpu_bench.cpp:52` and `spec-decode/engine/npu_spec_integration.cpp:26` both hardcode:
```
~/spec-decode/checkpoints/eagle3_draft_trained_420.bin
```
This is the checkpoint from the 343-sample/420-step run commit `2080abb3` already measured as inadequate (loss ~7.5 ≈ perplexity 1800, 0% acceptance — its own commit message calls this an "honest result"). Commit `6e469cec` (2026-07-12 05:03) later reconfigured training for a 5400-sample dataset (`target_cache_path` → `_v2`, ~3375 steps). On this machine, `~/checkpoints/1bit-spec/eagle3_qwen3_0.6b/` has step directories through `step_3370`/`step_latest`, and `~/spec-decode/checkpoints/eagle3_draft_v2.bin` exists (mtime Jul 12 13:28) — evidence a larger run completed. But commit `10750108` (2026-07-12 19:08, *after* `v2.bin` existed), which explicitly claims to "wire trained Eagle3 draft model... replacing naive early-exit" (closing #57), still points both C++ entry points at the **old** `_420` checkpoint.

Caveat: this evidence lives on the local machine (`~/spec-decode/`, `~/checkpoints/`), not in the repo — the repo's own committed `spec-decode/train_data/perfectblend_train*.jsonl` still shows only 360 lines despite commit messages claiming a 5400-sample dataset, so the training data behind `_v2` isn't checked in and this can't be fully verified from repo state alone. Flagging for a human check on whether `_v2` was ever actually evaluated, and if so, whether it should replace `_420` in the wired path.

## What's clean (verified, not just assumed)
- `docs/wiki/performance.md` and `site/benchmarks.json` currently and accurately report spec-decode as "unresolved," matching the real training history — this is the opposite of the historical fabrication pattern; published claims are honest. Both findings above are about latent/unused tooling and an unevaluated checkpoint, not live false claims.
- `spec-decode/scripts_local/export_draft_weights.py` and `regen_train_data.py` are genuine, functional (real `safetensors` reads, real `model.generate()` calls).

## Suggested fix
Either implement the real `draft_model.forward()` call in `eval_draft.py` or delete/clearly-mark it as non-functional scaffolding. Separately, get a human decision on whether `eagle3_draft_v2.bin` was evaluated and should be wired into the two C++ entry points in place of `_420`.

## Related
- #57 — closed by `10750108`, which per finding 2 above may not have actually wired in a trained/working checkpoint.

**Comment by bong-water-water-bong** [comment]
Triage: **Finding 1 is fixed** — `eval_draft.py` no longer proposes `torch.randint` random tokens; it calls `draft_model.forward(...)`, and PR #177 further corrects that loop (anchor preservation + `torch.cat`) per #146. **Finding 2 remains a human decision**: which checkpoint to wire (`_420` vs the on-machine `_v2`) can't be resolved from repo state — the `_v2` training data isn't checked in. The C++ path's hardcoded checkpoint is now env-overridable (`SPEC_DRAFT_CHECKPOINT`, #154). Keeping open for the checkpoint-evaluation call.

**Comment by bong-water-water-bong** [comment]
Finding 1 is resolved: `eval_draft.py` now calls the real `draft_model.forward(...)` in the decode loop (~line 159); the random path is only an explicit fallback (`Using random draft (baseline — expect 0% acceptance)`) when the checkpoint can't load. The README Quick Start no longer silently fabricates its measurement.

Finding 2 needs your call and can't be resolved from repo state: whether `eagle3_draft_v2.bin` was actually evaluated and should replace `_420` in the two wired C++ entry points. The `_v2` training data still isn't checked in. Leaving open for that decision.

**Comment by bong-water-water-bong** [comment]
Resolved — `eval_draft.py` now calls `draft_model.forward()` for real inference instead of `torch.randint()`. The old random-token path is replaced. A `"Using random draft"` message remains for a fallback baseline path.
